### PR TITLE
Make transient branch not fetchable for cloned project

### PR DESCRIPTION
**Pull request by LePtitDev**
Based on https://github.com/ubisoft/NGitLab/pull/380#discussion_r1096012621

### CODE REVIEW

File: NGitLab.Mock/MergeRequest.cs [hunk 1]
```diff
@@ -380,8 +381,13 @@ private void RefreshInternalState()
             _divergedCommitsCount = Project.Repository.ComputeDivergence(TargetBranchHeadCommit, commonCommit);
 
             // Determine if there are conflicts, by performing a rebase on a transient copy of the source branch
-            var transientBranchName = "transient/" + Guid.NewGuid().ToString("N");
-            Project.Repository.CreateBranch(transientBranchName, _consolidatedSourceBranch);
+            var smallBranchName = Guid.NewGuid().ToString("N");
+            var transientBranchName = "refs/transient/" + smallBranchName;
+
+            var branchFullPath = Path.Combine(Project.Repository.FullPath, ".git/refs/transient", smallBranchName);
```
Comment (meziantou) [inline NGitLab.Mock/MergeRequest.cs]: We are using Libgit2sharp in the repo. So, I think you can use `repo.Refs.Add(transientBranchName, ...)`. Make you can need to add an internal method in `Project.Repository` or expose a way to get the LibGit2Sharp repository object.

Comment (LePtitDev) [reply NGitLab.Mock/MergeRequest.cs]: I can add a ref with LibGit2Sharp but this ref is not usable for rebase. I tried an other way, I use the merge without committing to check conflicts and I abort operation after that. See : https://stackoverflow.com/a/501461

File: NGitLab.Mock/Repository.cs [hunk 2]
```diff
@@ -636,6 +636,43 @@ public bool Rebase(UserRef user, string sourceBranch, string targetBranch)
...
+            var signature = new Signature(user.UserName, user.Email, DateTimeOffset.UtcNow);
+            var options = new MergeOptions
+            {
+                CommitOnSuccess = false,
+                FastForwardStrategy = FastForwardStrategy.NoFastForward,
+            };
+
+            Commands.Checkout(repo, upstream);
+
+            var resetTip = upstream.Tip;
```
Comment (meziantou) [inline NGitLab.Mock/Repository.cs:663]: Why is this needed? As you don't create a commit, the `Tip` should not change?

File: NGitLab.Mock/Repository.cs [hunk 3]
```diff
@@ -636,6 +636,43 @@ public bool Rebase(UserRef user, string sourceBranch, string targetBranch)
...
+            if (string.IsNullOrEmpty(targetBranch))
+                throw new ArgumentException("Cannot be null or empty", nameof(targetBranch));
+
+            var repo = GetGitRepository();
+
+            var head = repo.Head;
+            var branch = repo.Branches[sourceBranch];
+            var upstream = repo.Branches[targetBranch];
+
+            _ = repo.Network.Remotes[targetBranch];
```
Comment (meziantou) [inline NGitLab.Mock/Repository.cs]: Is it needed?

File: NGitLab.Mock/Repository.cs [hunk 4]
```diff
@@ -636,6 +636,43 @@ public bool Rebase(UserRef user, string sourceBranch, string targetBranch)
...
+
+            Commands.Checkout(repo, upstream);
+
+            var resetTip = upstream.Tip;
+
+            var mergeResult = repo.Merge(branch.Tip, signature, options);
+            var result = mergeResult.Status == MergeStatus.Conflicts;
+
+            repo.Reset(ResetMode.Hard, resetTip);
+            Commands.Checkout(repo, head);
```
Comment (meziantou) [inline NGitLab.Mock/Repository.cs:669]: Do you think we can use `new CheckoutOptions { CheckoutModifiers = CheckoutModifiers.Force }` and remove the previous `Reset`? (@louis-z)

Comment (louis-z) [reply NGitLab.Mock/Repository.cs:669]: That sounds reasonable. It's worth a shot.

Comment (LePtitDev) [reply NGitLab.Mock/Repository.cs:669]: If I don't use `Reset`, the merging operation is not aborted

image

Comment (meziantou) [reply NGitLab.Mock/Repository.cs:669]: You are right
